### PR TITLE
Handle data fields correctly

### PR DESCRIPTION
**Pull request by nabuskey**
Fixes https://github.com/nimakaviani/managed-delivery-k8s-plugin/issues/50. 

- Decouple the data field from `K8sData`. The data field is now a `Map<String, Any?>?`
- With this, credentials resource spec is changed slightly. The spinnaker account name is now placed in metadata. 
- Remove the `strategy.spinnaker.io/versioned` annotation when generating diffs. 
- ConfigMaps created through clouddriver is versioned by default. We add the `strategy.spinnaker.io/versioned` annotation to disable it for now.  We may need to revisit this in the future to take advantage of versioned secrets and configmaps.

### CODE REVIEW

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt [hunk 1]
```diff
@@ -91,8 +98,8 @@ data class K8sCredentialManifest(
     override var kind: String?,
     override var metadata: Map<String, Any?>,
     override var spec: K8sSpec?,
-    override var data: K8sData? = null,
+    override var data: Map<String, Any?>? = null,
     override var status: Status? = null,
 ): K8sManifest(apiVersion, kind, metadata, spec, data, status) {
-    override fun name(): String = "${metadata[TYPE]}-${data?.account}"
+    override fun name(): String = "${metadata[TYPE]}-${metadata[CLOUDDRIVER_ACCOUNT]}"
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:104]: why do we need to change our own spec to support config map? account name under `data` appears to be nicer. even moving `git` under data is probably ok. any particular reason for not doing the following?

```suggestion
    override fun name(): String = "${metadata[TYPE]}-${data?.get("account")}"
```

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt [hunk 2]
```diff
@@ -82,16 +87,16 @@ class CredentialsResourceHandler(
...
             null,
-            data
+            data.toMap() as Map<String, Any?>?
         )
     }
 
-    override suspend fun current(resource: Resource<CredentialsResourceSpec>): K8sCredentialManifest? =
-        super.current(resource)?.let {
+    override suspend fun current(r: Resource<CredentialsResourceSpec>): K8sCredentialManifest? =
+        super.current(r)?.let {
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt:99]: why the change?

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt [hunk 3]
```diff
@@ -44,10 +44,12 @@ class K8sResourceHandler (
...
-                    val imageString = find(manifest.spec as MutableMap<String, Any?>, IMAGE) as String?
-                    imageString?.let { image ->
-                        log.info("Deployed artifact $image")
-                        notifyArtifactDeployed(r, getTag(image))
+                    if (!it.spec.isNullOrEmpty()) {
+                        val imageString = find(manifest.spec as MutableMap<String, Any?>, IMAGE) as String?
+                        imageString?.let { image ->
+                            log.info("Deployed artifact $image")
+                            notifyArtifactDeployed(r, getTag(image))
+                        }
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:52]: wouldn't the following have done it, rather than having to add the if block?

```
val imageString = find(manifest.spec ?: mutableMapOf()) as String?
```

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt [hunk 4]
```diff
@@ -57,14 +59,26 @@ class K8sResourceHandler (
...
+            val copy =  resource.spec.template.metadata.toMutableMap()
+            if (copy.containsKey("annotations")) {
+                val copyMap = (copy["annotations"] as Map<String, Any?>).toMutableMap()
+                copyMap["strategy.spinnaker.io/versioned"] = "false"
+                copy["annotations"] = copyMap
+            } else {
+                copy["annotations"] = mapOf("strategy.spinnaker.io/versioned" to "false")
+            }
+            resource.spec.template.metadata = copy
         }
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:81]: unclear about this bit. We ask for the annotation to be removed here only to put it back in this code snippet. what's the reason?

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt [hunk 5]
```diff
@@ -57,14 +59,26 @@ class K8sResourceHandler (
...
-        imageString?.let{
-            log.info("Deploying artifact $it")
-            notifyArtifactDeploying(resource, getTag(it))
+        if (!resource.spec.template.spec.isNullOrEmpty()) {
+            val imageString = find(resource.spec.template.spec as MutableMap<String, Any?>, "image") as String?
+            imageString?.let{
+                log.info("Deploying artifact $it")
+                notifyArtifactDeploying(resource, getTag(it))
+            }
+        }
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:69]: same as above for the if block. I know I wrote this bit, and i could have done a better job!
```
val imageString = find(manifest.spec ?: mutableMapOf()) as String?
```

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt [hunk 6]
```diff
@@ -171,6 +170,7 @@ abstract class GenericK8sResourceHandler <S: GenericK8sLocatable, R: K8sManifest
             "artifact.spinnaker.io/version",
             "moniker.spinnaker.io/application",
             "moniker.spinnaker.io/cluster",
+            "strategy.spinnaker.io/versioned",
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/GenericK8sResourceHandler.kt:173]: see the later comment in `K8sResourceHandler`

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt [hunk 7]
```diff
@@ -31,19 +32,23 @@ class CredentialsResourceHandler(
     private val encoder: Base64.Encoder = Base64.getMimeEncoder()
 
     public override suspend fun toResolvedType(resource: Resource<CredentialsResourceSpec>): K8sCredentialManifest {
-        if ((resource.spec.template.metadata[TYPE] as String?).isNullOrBlank()) {
-            throw CredResourceTypeMissing("missing \".metadata.type\" for the credential")
+        try {
+            require(!(resource.spec.template.metadata[TYPE] as String?).isNullOrEmpty()) {"missing \".metadata.type\" field for the credential"}
+            require(!(resource.spec.template.metadata[CLOUDDRIVER_ACCOUNT] as String?).isNullOrEmpty()) {"missing \".metadata.account\" field for the credential"}
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt:37]: probably should get rid of `field` in the msg. The field could be there and be empty.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt (hunk 2 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt:99]: So the super method's argument is named `r` and I think it's a good practice to match it in case it's called with named argument.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt (hunk 3 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:52]: Ah good idea. Will change

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt (hunk 1 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:104]: From end user's perspective, I think it's nicer to have users put information under one location, `metadata`, than have spread them out into metadata and data fields. Data fields are encoded and it's kind of pain to figure out where a particular secret came from. I think metadata is a better place to indicate users which clouddrvier account this secret  corresponds to. You can't really put your own fields in data through this anyways.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt (hunk 5 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:69]: Yep will change.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt (hunk 4 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:81]: When you are asking Clouddriver to deploy manifests, clouddriver will add the `strategy.spinnaker.io/versioned` annotation and version configMaps and secrets. If the annotation is present and set to false, it will not do that. 

During upsert, this annotation is added for the reason above. When you get the manifest from kubernetes for this resource during resource check, the annotation field contains the `strategy.spinnaker.io/versioned` key. When you do a diff against this and resource spec without removing the key, keel will never realize it's in sync because technically it's different. (The `last-applied-configuration` contains the annotation key while resource spec in keel doesn't)

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt (hunk 7 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt:37]: I'll update the message to indicate it's either empty or absent.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt (hunk 4 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:81]: Thinking about this, wonder if upsert is not the right place to do this. Upsert probably should not touch manifests being deployed. Should be done in `toResolvedType`?

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt (hunk 2 above)
Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt:99]: in this case since the method is small, it is probably not as bad. Generally though, with bigger methods, the abbreviated form reduces readability. This would have been frowned upon if it was done for a k8s project.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt (hunk 1 above)
Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:104]: hmmm, not sure I am following. what we have here under `K8sCredentialManifest` is what the users of MD see when they write their delivery config. It doesn't have much to do with the encoding that gets passed along as a secret. We take one `K8sCredentialManifest` and convert it to another `K8sCredentialManifest` [here](https://github.com/nimakaviani/managed-delivery-k8s-plugin/blob/b4404859614388e826080e0b139d48590da2eecd/k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt#L80-L89) but they are independent of one another.

I agree with the part that splitting the account type and the account name  between metadata and data section is probably not a good idea, but having both under metadata is a bit weird because what gets passed is actually account data IMO.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt [hunk 8]
```diff
@@ -10,8 +12,13 @@ data class K8sData(
...
     val password: String? = null,
-    val identity: String? = null
-)
+    val identity: String? = null,
+    val known_hosts: String? = null
+) {
+    fun toMap(): Map<String, String?> {
+        return jacksonObjectMapper().convertValue(this, object: TypeReference<Map<String, String?>>() {})
+    }
+}
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:21]: we can probably get rid of `K8sData` all together. It doesn't provide much value anymore.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt [hunk 9]
```diff
@@ -82,16 +87,16 @@ class CredentialsResourceHandler(
             SECRET,
             mapOf(
                 "namespace" to resource.spec.namespace,
-                "name" to "${resource.spec.template.metadata["type"]}-${resource.spec.template.data?.account}",
+                "name" to "${resource.spec.template.metadata["type"]}-${clouddriverAccountName}",
                 "annotations" to mapOf("strategy.spinnaker.io/versioned" to "false")
             ),
             null,
-            data
+            data.toMap() as Map<String, Any?>?
```
Comment (nimakaviani) [inline k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/CredentialsResourceHandler.kt:94]: if we remove `K8sData` this can be a simple map too so no converting back and forth.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt (hunk 4 above)
Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:81]: I think I came across this when I was writing the `cleanup` code but for some reason it wasn't a problem then. Do you remember why?

what do we lose if we add the `strategy.spinnaker.io/versioned: false` annotation to a resource for which clouddriver has had a versioning strategy? wouldn't it result in clouddriver not picking up changes and hence not reconciling things properly?

To your point, I think if we do it in `toResolvedType` we can modify desired, so we dont need to remove once and add it back again in two different places.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt (hunk 1 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:104]: For example, when you have to troubleshoot an issue with a secret created by this handler, wouldn't it be nicer to be able to just see where it comes from by just running `kubectl get secret example` and look at metadata? Instead of doing something like `{{index .data "username" | base64decode}}`?

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt (hunk 4 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:81]: Spinnaker adds that particular annotation for configmap and secrets only. This plugin hasn't worked for these resource types so it wasn't a problem. But now we are supporting these types, we are seeing this issue. 
If we do not annotate, the configMap looks something like this for a resource named `game-config`
```
NAME               DATA   AGE
game-config-v000   2      4d20h
```
This is fine as it is. Problem is with current interface for clouddriver k8s provider, there's no way to get this resource and keel will think the resource does not exist and tries to provision endlessly. 
The right thing to do is to find a way to get versioned resources through clouddriver and get the correct version. 
This is what I meant by:
```
ConfigMaps created through clouddriver is versioned by default. We add the strategy.spinnaker.io/versioned annotation to disable it for now. We may need to revisit this in the future to take advantage of versioned secrets and configmaps.
```
This is out of scope for this PR IMO

Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/resolver/K8sResourceHandler.kt:81]: yup sounds good.

File: k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt (hunk 1 above)
Comment (nabuskey) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:104]: Per our offline conversation: 
1. We will not change the spec
2. We will add annotations to resulting secrets to indicate where this secret was derived from. This however needs to be done after we implement our own annotation process.

Comment (nimakaviani) [reply k8s-plugin/src/main/kotlin/com/amazon/spinnaker/keel/k8s/model/K8sManifests.kt:104]: sounds good, thanks!

**Comment by nimakaviani** [comment]
really like the final outcome. moving the annotation to `toResolvedType` for `CredentialsResourceHandler` has made it much cleaner.

LGTM
